zerver/views/auth.py

```diff
@@ -357,10 +357,30 @@ def remote_user_sso(
     if not auth_enabled_helper([ZulipRemoteUserBackend.auth_backend_name], realm):
         return redirect_to_config_error("remoteuser/backend_disabled")
 
-    try:
-        remote_user = request.META["REMOTE_USER"]
-    except KeyError:
-        return redirect_to_config_error("remoteuser/remote_user_header_missing")
+    if not settings.REMOTE_USER_AUTH_USER_HEADER:
+        # Look for REMOTE_USER environment variable header
+        try:
+            remote_user = request.META["REMOTE_USER"]
+        except KeyError:
+            return redirect_to_config_error("remoteuser/remote_user_header_missing")
+    else:
+        user_header = settings.REMOTE_USER_AUTH_USER_HEADER
+        secret_header = settings.REMOTE_USER_AUTH_SECRET_HEADER
+        if secret_header:
+            # Check the shared secret header exists and contains the correct value
+            try:
+                if request.headers[secret_header] != settings.REMOTE_USER_AUTH_SECRET:
+                    logging.warning(f"SSO shared secret doesn't match value from {secret_header}")
+                    return redirect_to_config_error("remoteuser/remote_user_http_secret_incorrect")
+            except KeyError:
+                logging.warning(f"SSO request missing shared secret header: {secret_header}")
+                return redirect_to_config_error("remoteuser/remote_user_http_secret_missing")
+
+        logging.debug(f"Looking for SSO HEADER {user_header}")
+        try:
+            remote_user = request.headers[user_header]
+        except KeyError:
+            return redirect_to_config_error("remoteuser/remote_user_http_header_missing")
```

Comment:
I think it'd be better to extract the logic inside the conditional blocks into separate functions, to have the pattern here be something like

```suggestion
    if not settings.REMOTE_USER_AUTH_USER_HEADER:
        # Look for REMOTE_USER environment variable header
        remote_user, error_response = get_remote_user_from_env(request)
    else:
        remote_user, error_response = get_remote_user_from_headers(request)

    if error_response is not None:
        return error_response
```